pkg/client/introduce/client.go

```diff
@@ -0,0 +1,137 @@
+type provider interface {
+	Service(id string) (interface{}, error)
+	StorageProvider() storage.Provider
+}
+
+// Client enable access to introduce API
+type Client struct {
+	service.Event
+	service service.DIDComm
+	store   storage.Store
+}
+
+// New return new instance of introduce client
+func New(ctx provider) (*Client, error) {
+	svc, err := ctx.Service(introduce.Introduce)
+	if err != nil {
+		return nil, err
+	}
+
+	introduceSvc, ok := svc.(service.DIDComm)
+	if !ok {
+		return nil, errors.New("cast service to Introduce Service failed")
+	}
+
+	store, err := ctx.StorageProvider().OpenStore(introduce.Introduce)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		Event:   introduceSvc,
+		service: introduceSvc,
+		store:   store,
+	}, nil
+}
+
+func (c *Client) handleOutbound(msg interface{}) error {
+	payload, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	didMsg, err := service.NewDIDCommMsg(payload)
+	if err != nil {
+		return err
+	}
+
+	return c.service.HandleOutbound(didMsg, &service.Destination{})
+}
+
+// Options keeps the information needed for sending a proposal
+type Options struct {
+	// Invitation must be set when we have a public Invitation
+	Invitation *didexchange.Invitation
+	// Destinations contain one or two elements
+	// one element - for a public Invitation, otherwise two elements
+	Destinations []*service.Destination
+	// ThreadID must be an ID from incoming Request
+	ThreadID string
+}
+
+// SendProposal sends proposal
+// the function may receive an invitation or <nil> (Options.Invitation)
+// if the invitation is not <nil> it means we have a public invitation.
+// Destinations (Options.Destinations) should have one or two values, if destinations have only one value it means
+// that it is public invitation (an invitation `Options.Invitation` is required), otherwise if destinations have two
+// values, it means we do not have a public invitation (we will wait for it from one of introducees)
+// ThreadID (Options.ThreadID) need to be used when we received Request
+// Usage:
+//  Sends public invitation to the destination
+// 	SendProposal(&Options{
+// 		Invitation: invitation,
+// 		Destinations: []*service.Destination{
+// 			destination,
+// 		},
+// 	})
+// 	Forwards an invitation from one introducee to another
+// 	SendProposal(&Options{
+// 		Destinations: []*service.Destination{
+// 			destination1,
+// 			destination1,
+// 		},
+// 	})
+//  Sends public invitation to the destination (Request was received)
+// 	SendProposal(&Options{
+// 		Invitation: invitation,
+// 		Destinations: []*service.Destination{
+// 			destination,
+// 		},
+//		ThreadID: 8618e938-b68f-4712-b59c-f86f9a80d3c5,
+// 	})
+// 	Forwards an invitation from one introducee to another (Request was received)
+// 	SendProposal(&Options{
+// 		Destinations: []*service.Destination{
+// 			destination1,
+// 			destination1,
+// 		},
+//		ThreadID: 8618e938-b68f-4712-b59c-f86f9a80d3c5,
+// 	})
```

Comment:
@llorllale @troyronda @sandrask  This approach assumes that we have all data before the start state machine, but there is only one exception when we need an invitation from one of the introducees (it should be done through Continue(WithInvintation())), also after we received a request we should not execute Continue() (I am gonna make it panicked) instead of it we should SendProposal with threadID (threadID is equal to request.ID we received) we need this threadID to avoid doing Continue(WithInvintation()) I mean when we are sending request invitation can be saved to the DB

Summary:
* Continue() will be executed almost every time without args
* Continue(WithInvintation()) - will be executed only once by one of the introducees (if the introduction was started without request message)
* Continue() can not be executed after receiving the request